src/integration/java/io/pinecone/integration/dataPlane/UpsertErrorTest.java

```diff
@@ -0,0 +1,197 @@
+package io.pinecone.integration.dataPlane;
+
+import io.pinecone.clients.AsyncIndex;
+import io.pinecone.clients.Index;
+import io.pinecone.configs.PineconeConnection;
+import io.pinecone.exceptions.PineconeException;
+import io.pinecone.exceptions.PineconeValidationException;
+import io.pinecone.helpers.RandomStringBuilder;
+import io.pinecone.proto.VectorServiceGrpc;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+import static io.pinecone.helpers.BuildUpsertRequest.generateSparseIndicesByDimension;
+import static io.pinecone.helpers.BuildUpsertRequest.generateVectorValuesByDimension;
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class UpsertErrorTest {
+
+    private static Index index;
+    private static AsyncIndex asyncIndex;
+
+    @BeforeAll
+    public static void setUp() throws IOException, InterruptedException {
+        PineconeConnection connectionMock = mock(PineconeConnection.class);
+
+        VectorServiceGrpc.VectorServiceBlockingStub stubMock = mock(VectorServiceGrpc.VectorServiceBlockingStub.class);
+        VectorServiceGrpc.VectorServiceFutureStub asyncStubMock = mock(VectorServiceGrpc.VectorServiceFutureStub.class);
+
+        when(connectionMock.getBlockingStub()).thenReturn(stubMock);
+        when(connectionMock.getFutureStub()).thenReturn(asyncStubMock);
+
+        index = new Index(connectionMock);
+        asyncIndex = new AsyncIndex(connectionMock);
```

Comment:
Again, I'm confused why we're mocking here rather than just making the actual requests and validating the expected errors. Is this to speed up the tests?